jetty-server/src/main/java/org/eclipse/jetty/server/handler/HandlerCollection.java

```diff
@@ -86,44 +86,48 @@ public HandlerCollection(boolean mutableWhenRunning, Handler... handlers)
      */
     public void setHandlers(Handler[] handlers)
     {
-        if (!_mutableWhenRunning && isStarted())
-            throw new IllegalStateException(STARTED);
-
-        if (handlers!=null)
+        synchronized (this)
         {
-            // check for loops
-            for (Handler handler:handlers)
-                if (handler == this || (handler instanceof HandlerContainer &&
-                    Arrays.asList(((HandlerContainer)handler).getChildHandlers()).contains(this)))
-                        throw new IllegalStateException("setHandler loop");
-          
-            // Set server
-            for (Handler handler:handlers)
-                if (handler.getServer()!=getServer())
-                    handler.setServer(getServer());
+            if (!_mutableWhenRunning && isStarted())
+                throw new IllegalStateException(STARTED);
+
+            if (handlers!=null)
+            {
+                // check for loops
+                for (Handler handler:handlers)
+                    if (handler == this || (handler instanceof HandlerContainer &&
+                        Arrays.asList(((HandlerContainer)handler).getChildHandlers()).contains(this)))
+                            throw new IllegalStateException("setHandler loop");
+
+                // Set server
+                for (Handler handler:handlers)
+                    if (handler.getServer()!=getServer())
+                        handler.setServer(getServer());
+            }
+            Handler[] old=_handlers;
+            _handlers = handlers;
+            updateBeans(old, handlers);
         }
-        Handler[] old=_handlers;;
-        _handlers = handlers;
-        updateBeans(old, handlers);
     }
 
     /* ------------------------------------------------------------ */
-    /**
-     * @see Handler#handle(String, Request, HttpServletRequest, HttpServletResponse)
-     */
     @Override
     public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response)
         throws IOException, ServletException
     {
-        if (_handlers!=null && isStarted())
+        if (isStarted())
         {
+            Handler[] handlers = _handlers;
```

Comment:
no synchronized(this) on _handler access, is this safe?

Reply:
not needed because _handlers is a volatile